server/s3_list_object_versions_action.cc

```diff
@@ -0,0 +1,835 @@
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  if (motr_kv_reader->get_state() == S3MotrKVSReaderOpState::missing) {
+    s3_log(S3_LOG_DEBUG, request_id, "No versions found in listing\n");
+    // Reset state
+    check_any_keys_after_prefix = false;
+    fetch_successful = true;  // With no entries.
+  } else if (motr_kv_reader->get_state() ==
+             S3MotrKVSReaderOpState::failed_to_launch) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Bucket metadata next keyval operation failed due to pre launch "
+           "failure\n");
+    set_s3_error("ServiceUnavailable");
+  } else if (motr_kv_reader->get_state() ==
+             S3MotrKVSReaderOpState::failed_e2big) {
+    s3_log(S3_LOG_ERROR, request_id,
+           "Next keyval operation failed due rpc message size threshold\n");
+    if (max_record_count <= 1) {
+      set_s3_error("InternalError");
+      fetch_successful = false;
+      s3_log(S3_LOG_ERROR, request_id,
+             "Next keyval operation failed due rpc message size threshold, "
+             "even with max record retrieval as one\n");
+    } else {
+      retry_count++;
+      s3_log(S3_LOG_INFO, request_id, "Retry next key val, retry count = %d\n",
+             retry_count);
+      get_next_versions();
+      s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+      return;
+    }
+  } else {
+    s3_log(S3_LOG_DEBUG, request_id, "Failed to find versions listing\n");
+    set_s3_error("InternalError");
+    fetch_successful = false;
+  }
+  send_response_to_s3_client();
+  s3_log(S3_LOG_DEBUG, "", "%s Exit", __func__);
+}
+
+void S3ListObjectVersionsAction::check_latest_versions() {
+  s3_log(S3_LOG_INFO, stripped_request_id, "%s Entry\n", __func__);
+  std::shared_ptr<S3ObjectMetadata> version_metadata = nullptr;
+  size_t version_list_offset_temp = 0;
+  // Run through version_list, check and set "latest" flag.
+  for (auto&& version : versions_list) {
+    if (last_object_checked.empty()) {
+      version_metadata = version;
+      break;
+    } else if ((version_list_offset_temp <= version_list_offset) ||
+               (last_object_checked == version->get_object_name())) {
+      version_list_offset_temp++;
+      continue;
+    } else {
+      version_metadata = version;
+      break;
+    }
+  }
+  if (version_metadata != nullptr) {
+    version_list_offset = version_list_offset_temp;
+    last_object_checked = version_metadata->get_object_name();
+    const auto& object_list_idx_lo =
+        bucket_metadata->get_object_list_index_layout();
+    const auto& obj_version_list_idx_lo =
+        bucket_metadata->get_objects_version_list_index_layout();
+
+    if (zero(object_list_idx_lo.oid) || zero(obj_version_list_idx_lo.oid)) {
+      // Object list index and version list index missing.
+      get_next_versions_failed();
+    } else {
+      // Read object metadata from object index.
+      object_metadata = object_metadata_factory->create_object_metadata_obj(
```

Comment:
I don't see point in fetching latest `object_metadata` from the object list index for every iteration as we anyway have local copy of version objects in `version_list`.  
By doing we may have a situation where we don't find the latest object at all in the `version_list` as the latest object version might get updated during this time and not be present in the local copy of `version_list`. ​
Also fetching latest object_metadata for every entry in version_list is costly.
We can simply fetch the latest object_metadata from the object list index once and compare the latest version with what we have in version_list and break the loop when we find a match.

Reply:
We do fetch the object_metadata from object list only once per object. We do not fetch the same for every version of that object in the version_list. Please check the logic in "for" loop above.

"By doing we may have a situation where we don't find the latest object at all in the version_list as the latest object version might get updated during this time and not be present in the local copy of version_list. "
I guess this situation will be there for the time till the version_list and object list are in sync.

Reply:
@sjain09 As discussed, above pending comment will be handled in https://jts.seagate.com/browse/EOS-26911

Reply:
Will be addressed in jira : https://jts.seagate.com/browse/EOS-27169